File: frontend/src/api/employees.ts
import apiClient from "@/lib/apiClient";
import type {
  Employee,
  EmployeeCreate,
  EmployeeListParams,
  EmployeePatch,
  EmployeeUpdate,
  PaginatedResponse,
} from "@/types/employee";

export async function listEmployees(
  params: EmployeeListParams = {},
): Promise<PaginatedResponse<Employee>> {
  const query = {
    page: params.page ?? 1,
    page_size: params.page_size ?? 20,
    q: params.search ?? "",
    country: params.country ?? "",
    department: params.department ?? "",
    job_title: params.job_title ?? "",
    status: params.status ?? "",
    sort_by: params.sort_by ?? "full_name",
    sort_order: params.sort_order ?? "asc",
  };
  const { data } = await apiClient.get<PaginatedResponse<Employee>>("/employees", { params: query });
  return data;
}

export async function getEmployee(id: number): Promise<Employee> {
  const { data } = await apiClient.get<Employee>(`/employees/${id}`);
  return data;
}

export async function createEmployee(payload: EmployeeCreate): Promise<Employee> {
  const { data } = await apiClient.post<Employee>("/employees", payload);
  return data;
}

export async function updateEmployee(id: number, payload: EmployeeUpdate): Promise<Employee> {
  const { data } = await apiClient.put<Employee>(`/employees/${id}`, payload);
  return data;
}

export async function patchEmployee(id: number, payload: EmployeePatch): Promise<Employee> {
  const { data } = await apiClient.patch<Employee>(`/employees/${id}`, payload);
  return data;
}

export async function deleteEmployee(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.delete<{ message: string }>(`/employees/${id}`);
  return data;
}

File: backend/app/routers/employees.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.employee import EmployeeCreate, EmployeePatch, EmployeeResponse, EmployeeUpdate
from app.services.employee_service import (
    create_employee,
    delete_employee,
    get_employee_or_404,
    list_employees,
    patch_employee,
    update_employee,
)

router = APIRouter(prefix="/employees", tags=["employees"])


@router.get("/")
def list_employees_endpoint(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1),
    q: str = Query(default=""),
    country: str = Query(default=""),
    department: str = Query(default=""),
    job_title: str = Query(default=""),
    status: str = Query(default=""),
    employment_type: str = Query(default=""),
    sort_by: str = Query(default="full_name"),
    sort_order: str = Query(default="asc"),
    db: Session = Depends(get_db),
):
    return list_employees(
        db=db,
        page=page,
        page_size=page_size,
        search=q,
        country=country,
        department=department,
        job_title=job_title,
        status=status,
        sort_by=sort_by,
        sort_order=sort_order,
        employment_type=employment_type,
    )


@router.post("/", response_model=EmployeeResponse, status_code=201)
def create_employee_endpoint(payload: EmployeeCreate, db: Session = Depends(get_db)):
    return create_employee(db, payload)


@router.get("/{employee_id}", response_model=EmployeeResponse)
def get_employee_endpoint(employee_id: int, db: Session = Depends(get_db)):
    return get_employee_or_404(db, employee_id)


@router.put("/{employee_id}", response_model=EmployeeResponse)
def update_employee_endpoint(employee_id: int, payload: EmployeeUpdate, db: Session = Depends(get_db)):
    return update_employee(db, employee_id, payload)


@router.patch("/{employee_id}", response_model=EmployeeResponse)
def patch_employee_endpoint(employee_id: int, payload: EmployeePatch, db: Session = Depends(get_db)):
    return patch_employee(db, employee_id, payload)


@router.delete("/{employee_id}")
def delete_employee_endpoint(employee_id: int, db: Session = Depends(get_db)):
    return delete_employee(db, employee_id)
